documentation/specs/Modal.md

```diff
@@ -0,0 +1,191 @@
+# `Modal` Component Specification
+
+## Overview
+
+A `Modal` is a page overlay that displays information and blocks interaction with the page until an action is taken or the `Modal` is dismissed.
+
+## Design
+
+`Modal` will use `useDialog` and `useModalOverlay` from `react-aria` to provide the technical foundation for an accessible dialog.
+
+The component design was inspired both by Shopify's Modal component and Twilio's Modal component.
+
+`Modal` will be a controlled component deferring the state management to the consumer.
+
+`Modal` will be a compound component consistenting of `Modal`, `Modal.Header`, `Modal.Body`, and `Modal.Footer`.
+
+### API
+
+```ts
+type HeaderElementType = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
+
+type ModalProps = {
+  /**
+   * Whether the modal is open or not.
+   */
+  isOpen: boolean;
+
+  /**
+   * Callback when the modal is closed.
+   */
+  onClose: () => void;
+
+  /**
+   * Whether or not the modal can be dismissed.
+   *
+   * @default true
+   */
+  isDismissable?: boolean;
+};
+
+type ModalHeaderProps = {
+  /**
+   * Modal header element type. Should be a valid document heading level.
+   *
+   * @default "h2"
+   */
+  as?: HeaderElementType;
+
+  /**
+   * The content for the title of the modal.
+   */
+  children: string;
+
+  /**
+   * The content for the subtitle of the modal.
+   */
+  subtitle?: string;
+
+  /**
+   * Icon to display at the start of the header title.
+   */
+  iconAtStart?: {
+    accessibilityLabel?: string;
+    symbol: IconSymbol;
+  };
```

Comment:
i don't love having complex props like this but i couldn't think of a cleaner way to handle these slots that need to be flexible but only to an extent. inspired by Polaris' Modal who does similar for some of the slots. i'm open to better ideas.